### ISSUE TITLE
Subtitle shadow angle does not match with mpv

### ISSUE THREAD

**Issue by rasitayaz** [iina/iina#4770]
**What you want IINA to do:**

<img width="559" alt="image" src="https://github.com/iina/iina/assets/45076094/4a0788fe-8204-49e8-94cd-f1ed58888a2c">

Under 'text subtitles > advanced > shadow', it would be nice to have the option to customize the angle of the shadow.

**What IINA does currently:**

Displays subtitle shadows at a fixed angle.

**Why you think this should be added:**

I believe default shadow angle is too close to horizontal axis and it makes subtitles a bit harder to read.

<img width="243" alt="image" src="https://github.com/iina/iina/assets/45076094/dbc21f25-33a2-42d7-8ddd-1a789ab8a09f">

**Examples of other projects that have something similar:**
In VLC advanced settings;
<img width="887" alt="image" src="https://github.com/iina/iina/assets/45076094/9d2daac1-ceae-4d0d-8a40-0e3ec3d178a8">

**Comment by low-batt** [comment]
For audio/video and subtitle rendering support IINA uses a library from the [mpv](https://github.com/mpv-player/mpv) project. It is `mpv` that adds the subtitles to a video frame.

When it comes to changing how the subtitles are rendered IINA can only alter that using the options exposed by `mpv`. Those options are listed in the [Subtitles](https://mpv.io/manual/master/#subtitles) section of the `mpv` manual.

I looked through the available options and did not spot anything that could change the angle of subtitle shadows. Before IINA can offer UI controls for this the [mpv](https://github.com/mpv-player/mpv) project needs to provide an option to control this. A feature request needs to be entered with that project.

**Comment by rasitayaz** [comment]
You're right about mpv not having this option, but if you take a look at these examples;

IINA:
<img width="603" alt="image" src="https://github.com/iina/iina/assets/45076094/a8b4c17c-8bae-4691-b37f-a0e8c2250a55">

mpv player:
<img width="560" alt="image" src="https://github.com/iina/iina/assets/45076094/2c26132f-bc46-462f-8ae3-2febabc8ad55">

Despite having the same configuration (shadow offset: 10), these shadows are noticeably cast at different angles for IINA and the mpv player.

Upon further investigation, I attempted to approximate the angles. The shadow angle for mpv is approximately -45 degrees, whereas IINA's shadow is around -27 degrees, closely resembling the angle of a right triangle with sides x, 2x, and x√5. Bear with me here, I'm getting somewhere with this. :D

I've concluded that these shadows are not created based on an angle but rather a displacement. For instance, mpv shifts the shadow by 1 unit per offset both horizontally and vertically, while IINA shifts it 2 units horizontally and 1 unit vertically.

In summary, all I'm attempting to achieve is subtitle shadows at -45 degrees, which is a standard for most video players, including mpv apparently.

**Comment by low-batt** [comment]
The posted screenshots of subtitle rendering indicate something is wrong. Subtitle rendering should match. I have changed the labels applied to this issue to indicate a regression instead of a feature request.

I took a quick look into this. I'm thinking this is the problem described in:
https://github.com/mpv-player/mpv/issues/11825

That regression appeared when [libass](https://github.com/libass/libass) released version 0.17.0. This was reported to the `libass` project in: https://github.com/libass/libass/issues/690

The outcome of that was the change in `mpv` PR https://github.com/mpv-player/mpv/pull/11831 which was merged on Jul 16, 2023.

Checking an IINA 1.3.4 `mpv` log file shows the version of `mpv` being used by IINA was built before the above fix was merged and that the 0.17.x version of `libass` is being used:
```text
[   0.004][v][cplayer] mpv 0.35.0-419-gf79458476b-dirty Copyright © 2000-2023 mpv/MPlayer/mplayer2 projects
[   0.004][v][cplayer]  built on Tue May 30 08:54:21 2023
…
[   0.054][d][osd/libass] ASS library version: 0x1701000 (runtime 0x1701000)
[   0.054][v][osd/libass] libass API version: 0x1701000
[   0.054][v][osd/libass] libass source: tarball: 0.17.1
```

Do you think the problem described in https://github.com/mpv-player/mpv/issues/11825 matches the behavior you are experiencing?

**Comment by rasitayaz** [comment]
> Do you think the problem described in https://github.com/mpv-player/mpv/issues/11825 matches the behavior you are experiencing?

Doesn't quite look like the same, because my issue is related to shadows and the issue described in the link seems to be related to borders. But the same bug might be causing both of these issues.

Anyway, if you think that the subtitle styles should match, this will probably be fixed when IINA uses a newer mpv version. Thanks for the explanation and for your time.

I've also updated the issue title to match the labels.

**Comment by low-batt** [comment]
Thanks for updating the title. As we are unsure if this is the `mpv` issue identified I am going to take a closer look. I believe I have reproduced it. I will investigate.

**Comment by low-batt** [comment]
I invoked `mpv` with these options for the test:
```text
mpv --pause --start=1:09:52 --sub-shadow-color=#000000 --sub-shadow-offset=10
```

And configured IINA with similar options.

mpv 0.37.0:
![issue-4770-mpv](https://github.com/iina/iina/assets/86170219/eb190913-2635-4d2a-a4e6-8ac19c98ab59)

IINA 1.3.4:
![issue-4770-IINA](https://github.com/iina/iina/assets/86170219/6aa6959c-321b-4c2d-895f-9c8575ef07f9)

IINA with newer `libmpv`:
![issue-4770-IINA-new-mpv](https://github.com/iina/iina/assets/86170219/e1711c28-a247-4103-8b1b-12d217b75273)

The difference in the subtitle shadow is very apparent between "is" and "emptying".

I'm thinking the problem described in `mpv` issue https://github.com/mpv-player/mpv/issues/11825 is also causing this subtitle shadow behavior.

It is likely the next IINA release will be a follow up to 1.3.4 containing the fixes that were planed for 1.3.4 but weren't ready in time for the end of year release. It is unlikely that release will contain an upgraded `libmpv` although I will bring this issue up with the team. Pretty sure dependencies will be upgraded for the release after that one.

Thank you very much for pointing this problem out.

**Comment by low-batt** [comment]
The [mpv](https://mpv.io/) issue had to do with [libass](https://github.com/libass/libass) issue https://github.com/libass/libass/issues/690 concerning a change made for `libass` 0.17.0. Changes have been merged and `mpv` is now compatible with `libass` 0.17.0.

This issue prompted the logging of the `libass` version to IINA:
```text
16:19:07.754 [iina][d] Using mpv 0.35.0-419-gf79458476b-dirty and libass 0.17.1
```

The planned `mpv` upgrade should fix this issue.

**Comment by low-batt** [comment]
This was fixed in IINA 1.4.0-beta1.
